src/Oscoin/Crypto/Address.hs

```diff
@@ -0,0 +1,111 @@
+{- | An address format for the Oscoin blockchain loosely based on Parity's SS58
+-- and on BTC's Base58Check.
+--
+-- See docs/address_format.md for an extensive documentation.
+--
+--}
+
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Oscoin.Crypto.Address
+    ( Address(addressPrefix, addressPayload)
+    , AddressPrefix
+    , AddressPayload
+
+    -- * Creating an Address
+    , fromPublicKey
+
+    -- * Parsing binary blobs back into addresses
+    , parseAddress
+
+    -- * Pretty-printing an Address
+    , renderAddress
+    , fmtAddress
+    ) where
+
+import           Oscoin.Prelude
+
+import qualified Oscoin.Configuration as Config
+import           Oscoin.Crypto.Address.Internal
+import           Oscoin.Crypto.Address.Serialisation as Serialisation
+import           Oscoin.Crypto.Hash
+import           Oscoin.Crypto.PubKey
+
+import qualified Codec.CBOR.ByteArray as CBOR
+import qualified Codec.Serialise as CBOR
+import           Data.ByteArray (ByteArrayAccess, convert)
+import           Data.ByteString.BaseN
+                 (decodeBase32z, encodeBase32z, encodedText)
+import           Data.Tagged
+import qualified Formatting as F
+
+{------------------------------------------------------------------------------
+    Creating an Address
+------------------------------------------------------------------------------}
+
+data AddressCreationError =
+    UnsupportedConfiguration (Config.Network, Config.Environment)
+    deriving Show
+
+-- | NOTE(adn) It's conceivable that we might want to crate a type to carry
+-- around the context needed to create an 'Address', but for now single
+-- arguments should do.
+fromPublicKey
+    :: forall c.
+       ( ByteArrayAccess (ShortHash c)
+       , Hashable c (PublicKey c)
+       )
+    => Config.Network
+    -- ^ The 'Network' the node is running on.
+    -> Config.Environment
```

Comment:
Having to pass an environment every time you want to generate an address may become a problem, because now we have to thread it everywhere. I guess one option is to partially apply this function and store the partially-applied version somewhere in the node ... thoughts?

Reply:
The way I see this module, it's that it's a low-level one, which exposes just the building blocks you need to construct more abstractions on top of it. As such, we can surely do the "partially applied" trick elsewhere, methinks. Wdyt?

Reply:
Yeah sounds good.